### PR TITLE
F/search package

### PR DESCRIPTION
**Pull request by pranavkulkarni**
Module to search for ArcGIS items and format them for display in ArcGIS Hub

Structure --
```
src
  ago
    helpers
      aggs
      filters
      format
      page
      sort
    search.ts // searches for items and returns raw ago results
    params.ts // defines ISearchParams interface to hub.js/search
    serialize.ts // serialize raw UI params into query parameters that V3 understands
    format-item-collection.ts // format raw ago results into hub v3 style datasets
    compute-items-facets.ts // compute aggs/facets based on ago results and also calculate aggs for custom facets
```

Sample of how this will be used in UI code --

```
const authentication = new arcgisRest.UserSession({
      portal: 'https://qaext.arcgis.com/sharing/rest',
      username: 'dcadminqa',
      password: 'dcadminqa1'
    })
    const portal = 'https://qaext.arcgis.com/sharing/rest'
    const params = {
      q: 'crime',
      sort: 'name',
      agg: { fields: 'tags,collection,owner,source,hasApi,downloadable' },
      start: 1,
      num: 10
    }
    authentication.getToken(portal)
    .then(token => {
      const headers = { authorization: token, portal }
      const serializedParams = arcgisHub.serialize(params)
      console.log(serializedParams)
      return fetch(`hub.arcgis.com/api/v3/search?${serializedParams}`, { headers })
    })
    .then(unifiedResultsPromise => {
      return unifiedResultsPromise.json()
    })
    .then(unifiedResults => {
      console.log(unifiedResults)
      // UPDATE UI model with unified results
    })
```

Sample of how hub indexer api v3 new search route will use hub.js --

```
   const search = require('@esri/hub-search')

  // 1. perform ago search with facets and formatting
  const agoResults = await search.agoSearch(req.query, req.headers.authorization, req.headers.portal)

  // 2. compute items facets
  const facets = await search.computeItemsFacets(agoResults.aggregations, req.query, req.headers.authorization, req.headers.portal)

  // 3. format item collection based on actual search results and computed facets
  const model = search.agoFormatItemCollection(agoResults, facets, req.query)

  return model
```

### CODE REVIEW

File: test.html [hunk 1]
```diff
@@ -0,0 +1,78 @@
+<!DOCTYPE html>
```
Comment (jgravois) [inline test.html]: we don't want to check this one in.

File: packages/search/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+/* Copyright (c) 2018 Environmental Systems Research Institute, Inc.
```
Comment (jgravois) [inline packages/search/src/index.ts]: `2019`.

File: packages/search/src/ago/params.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import { IRequestOptions } from "@esri/arcgis-rest-request";
+
+/* Copyright (c) 2017 Environmental Systems Research Institute, Inc.
```
Comment (jgravois) [inline packages/search/src/ago/params.ts]: `2017` > `2019` (and move to the top of the file).

File: packages/search/src/ago/params.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+* Apache-2.0 */
+
+export interface ISearchParams extends IRequestOptions {
```
Comment (jgravois) [inline packages/search/src/ago/params.ts]: `ISearchOptions` (see https://github.com/Esri/hub.js/pull/152#discussion_r275059511)

File: packages/search/src/ago/helpers/filters/downloadable.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    typeKeywordFilter = downloadableTypeKeywords.map(type => {
+      return `-typekeywords:"${type}"`;
+    });
+  }
+  return `(${downloadFilter.concat(typeKeywordFilter).join(" OR ")})`;
+}
+
+const downloadableTypeKeywords = ["Data"];
+// eligible types are listed here: http://doc.arcgis.com/en/arcgis-online/reference/supported-items.htm
+const downloadableTypes = [
```
Comment (jgravois) [inline packages/search/src/ago/helpers/filters/downloadable.ts]: whether you keep this type in `common` or `search`, lets keep it DRY.

File: packages/search/src/ago/helpers/aggs/downloadable.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        formattedAggs.downloadable.false++;
+      }
+      return formattedAggs;
+    },
+    { downloadable: { true: 0, false: 0 } }
+  );
+}
+
+// eligible types are listed here: http://doc.arcgis.com/en/arcgis-online/reference/supported-items.htm
+const downloadableTypes = [
```
Comment (jgravois) [inline packages/search/src/ago/helpers/aggs/downloadable.ts]: given what we already started in `@esri/hub-common`, would it make more sense to export this type from there?

File: packages/search/src/ago/compute-items-facets.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+};
+
+/*
+ * This Util function takes AGO results and formats them into an aggregations object
+ * that looks like it came from the V3 API. Many facets can be fairly easily mocked
+ * using a standardized approach (the 'else if' and 'else' branches) but a few facets
+ * require more complex and customized logic. Those go in the `customFacets` hash,
+ * where the name of the key is the name of the facet being computed and the custom function
+ * is implemented below.
+ */
```
Comment (jgravois) [inline packages/search/src/ago/compute-items-facets.ts]: this code comment isn't being picked up by TSDoc.

Screen Shot 2019-05-01 at 9 20 25 AM

for an example of method documentation that's formatted for display in our documentation, see: https://github.com/Esri/hub.js/blob/b336f324b1639993309ead0c87f91b13f737d97e/packages/annotations/src/add.ts#L29-L50

File: packages/search/src/ago/encode-ago-query.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
+import { getSortField } from "./helpers/sort";
+import { createFilters, handleFilter } from "./helpers/filters";
+import { createAggs } from "./helpers/aggs";
+
+/**
+ * Construct a query object with filters, and queryParams sent by hub indexer
+ * @param queryObject any
+ */
+export function encodeAgoQuery(queryParams: any = {}) {
```
Comment (jgravois) [inline packages/search/src/ago/encode-ago-query.ts]: all `export`ed functions show up in the documentation unless they are added to the explicit blacklist below.

i think it'd be appropriate to hide a lot of your shared utility functions, but i'll leave it up to you to decide exactly which.

https://github.com/Esri/hub.js/blob/da48fbc9e13ce46223086cdb9082a39fd4128d18/docs/build-typedoc.js#L138

```ts
const blacklist = [ "encodeAgoQuery", "etc" ];
```

File: packages/search/package.json [hunk 9]
```diff
@@ -0,0 +1,61 @@
+{
+  "name": "@esri/hub-search",
+  "version": "1.10.0",
+  "description": "Module to search for ArcGIS items and format them for display in ArcGIS Hub.",
```
[the remark this answers is not in the document]
Comment (pranavkulkarni) [reply packages/search/package.json]: actually, we will be using this on front end too.

File: test.html (hunk 1 above)
Comment (pranavkulkarni) [reply test.html]: It is there in .gitignore. Not sure why it has appeared 😕

File: packages/search/src/ago/compute-items-facets.ts (hunk 7 above)
Comment (pranavkulkarni) [reply packages/search/src/ago/compute-items-facets.ts]: nice! I will fix that

File: test.html (hunk 1 above)
Comment (jgravois) [reply test.html]: i think you overrode the .gitignore at some point with an explicit `git add test.html`.

luckily you can `git rm test.html` too.

**Comment by jgravois** [comment]
now that v2.0.0 has shipped, we'll land this from #152

**Pull request closed without merging**